### DISCUSSION TITLE
Bringing back RDFa support to RDFLib

### DISCUSSION THREAD

**Discussion by fathoni** [RDFLib/rdflib#1582] [Development]
RDFa support was [dropped ](https://github.com/RDFLib/rdflib/pull/828) from RDFLib since RDFLib version 5.0. There have been changes to the RDFLib 5.0 which breaks its compatibility with pyRDFa3, and unfortunately, there is currently no plan on maintaining its compatibility according to the [discussion](https://github.com/RDFLib/pyrdfa3/issues/33) on pyRDFa3's GitHub page.  I think It would be great to bring RDFa compatibility back to the current version.  Thank you!

**Comment by mwatts15** [comment]
Can we describe what the incompatibilities are with RDFLib 5+? The only problem I see in the linked issue is that `file` is no-longer a built-in or a type in Python 3, so there's a reference error (also, pyRrdfa3 is severely lacking in error reporting...though that's a separate issue).

**Reply by ghost** [reply]
> Can we describe what the incompatibilities are with RDFLib 5+? The only problem I see in the linked issue is that `file` is no-longer a built-in or a type in Python 3, so there's a reference error (also, pyRrdfa3 is severely lacking in error reporting...though that's a separate issue).


I haven't been through it exhaustively but bringing the RDFLib parsers up to date was fairly straightforward, the basic functionality seems okay. I have a [restore-pyrdfa](https://github.com/gjhiggins/rdflib/tree/restore-pyrdfa) branch which you might like to check out and play with in a venv.

I used the following rather undemanding test:

```py
g = Graph().parse(location="https://alistapart.com/article/introduction-to-rdfa/", format="text/html")
g.serialize(format="ttl")
```
```
@prefix dc1: <http://purl.org/dc/terms/> .
@prefix ns1: <article:> .
@prefix ns2: <http://ogp.me/ns#image:> .
@prefix ns3: <spp:> .
@prefix ns4: <http://www.w3.org/1999/xhtml/vocab#> .
@prefix og: <http://ogp.me/ns#> .
@prefix schema: <https://schema.org/> .
@prefix schema1: <http://schema.org/> .

<https://alistapart.com/article/introduction-to-rdfa/> ns1:modified_time "2009-06-23T14:00:00+00:00"@en-us ;
    ns1:published_time "2009-06-23T14:00:00+00:00"@en-us ;
    og:description "In part one of a two-part primer on RDFa, learn how semantic features normally confined to the head of an HTML document can be used to add semantic richness to the elements of the body. Mark Birbec…"@en-us ;
    og:image "https://i0.wp.com/alistapart.com/wp-content/uploads/2012/07/RDFa.jpg?fit=540%2C221&ssl=1"@en-us ;
    ns2:alt "Introduction to RDFa"@en-us ;
    ns2:height "221"@en-us ;
    ns2:width "540"@en-us ;
    og:locale "en_US"@en-us ;
    og:site_name "A List Apart"@en-us ;
    og:title "Introduction to RDFa"@en-us ;
    og:type "article"@en-us ;
    og:url "https://alistapart.com/article/introduction-to-rdfa/"@en-us ;
    dc1:creator "..."@en-us ;
    dc1:issued "2009-06-23"@en-us ;
    <https://api.w.org/> <https://alistapart.com/wp-json/> ;
    ns3:photo-attribution <https://alistapart.com/article/introduction-to-rdfa/nytimes.com> .

[] a schema:WebPage ;
    schema:copyrightHolder " A List Apart & Our Authors"@en-US ;
    schema:copyrightYear " 1998–2021"@en-US ;
    schema:url <http://abookapart.com/>,
        <http://aneventapart.com/>,
        <https://alistapart.com/about/copyright/>,
        <https://alistapart.com/about/privacy-policy/>,
        <https://webtype.com/> .

[] a schema1:SiteNavigationElement ;
    schema1:name "Articles"@en-US,
        "Events"@en-US,
        "Home"@en-US,
        "Italian"@en-US,
        "Language (English) "@en-US,
        "Topics"@en-US,
        "Write for Us"@en-US ;
    schema1:url <https://alistapart.com>,
        <https://alistapart.com/about/contribute/>,
        <https://alistapart.com/articles/>,
        <https://alistapart.com/events/>,
        <https://alistapart.com/it>,
        <https://alistapart.com/topics/> .

[] ns4:role ns4:search .

[] a schema1:Article ;
    schema1:author [ a schema1:Person ;
            schema1:name "Tom Greenwood"@en-US ;
            schema1:url <https://alistapart.com/author/tom-greenwood/> ],
        [ a schema1:Person ;
            schema1:name "Mark Birbeck"@en-US ;
            schema1:url <https://alistapart.com/author/mbirbeck/> ],
        [ a schema1:Person ;
            schema1:name "Sophia V. Prater"@en-US ;
            schema1:url <https://alistapart.com/author/sophiavoychehovski/> ],
        [ a schema1:Person ;
            schema1:name "Mike Wills"@en-US ;
            schema1:url <https://alistapart.com/author/mike-wills/> ],
        [ a schema1:Person ;
            schema1:name "Eva PenzeyMoog"@en-US ;
            schema1:url <https://alistapart.com/author/eva-penzeymoog/> ],
        [ a schema1:Person ;
            schema1:name "Patrick Brosset"@en-US ;
            schema1:url <https://alistapart.com/author/patrick-brosset/> ] ;
    schema1:headline "Introduction to RDFa"@en-US ;
    schema1:image [ a schema1:ImageObject ;
            schema1:contentUrl <https://149572954.v2.pressablecdn.com/wp-content/uploads/2012/07/RDFa.jpg> ;
            schema1:issueNumber "286"@en-US ] .

[] a schema1:ImageObject .

[] a schema1:SiteNavigationElement ;
    schema1:url <https://alistapart.com/about/>,
        <https://alistapart.com/about/masthead/>,
        <https://alistapart.com/about/rss-feed/>,
        <https://alistapart.com/about/sponsor/>,
        <https://alistapart.com/about/style-guide/>,
        <https://alistapart.com/authors/>,
        <https://alistapart.com/contact/>,
        <https://alistapart.com/email-signup/>,
        <https://www.facebook.com/alistapart>,
        <https://www.twitter.com/alistapart> .
```

**Reply by mwatts15** [reply]
I see. Just going from the master branch of pyRDFa3 or the version from PyPI, a number of name reference errors are encountered. It looks like very sloppy work setting up pyRDFa3 in that repo.

**Reply by ghost** [reply]
> I see. Just going from the master branch of pyRDFa3 or the version from PyPI, a number of name reference errors are encountered. It looks like very sloppy work setting up pyRDFa3 in that repo.

Could you be specific about the process in which you are encountering name reference errors? All I'm doing (afaik) is installing pyrdfa3 from PyPi. I did copy over pymicroformats because atm there's no PyPi installation route but that can be elided by specifying the repos details in requirements.txt (not sure how to do the same in setup.py)

**Reply by mwatts15** [reply]
Yes. I created a virtualenv, cloned pyrdfa3, installed pyrdfa3 in the virtualenv and then ran your code snippet. It failed with
```
Traceback (most recent call last):
  File "/home/markw/work/pyrdfa3/env/lib/python3.9/site-packages/rdflib/plugin.py", line 124, in get
    p: Plugin[PluginT] = _plugins[(name, kind)]
KeyError: ('text/html', <class 'rdflib.parser.Parser'>)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/markw/work/pyrdfa3/temp/discussion-1582.py", line 8, in <module>
    g = Graph().parse(location="https://alistapart.com/article/introduction-to-rdfa/", format="text/html")
  File "/home/markw/work/pyrdfa3/env/lib/python3.9/site-packages/rdflib/graph.py", line 1255, in parse
    parser = plugin.get(format, Parser)()
  File "/home/markw/work/pyrdfa3/env/lib/python3.9/site-packages/rdflib/plugin.py", line 126, in get
    raise PluginException("No plugin registered for (%s, %s)" % (name, kind))
rdflib.plugin.PluginException: No plugin registered
```
I saw that the entry points were both malformed and missing `text/html` (my comments added):
```
        'rdf.plugins.parser': [
            'rdfa = pyRdfa.rdflibparsers:RDFaParser',
            'rdfa1.0 = pyRdfa.rdflibparsers:RDFa10Parser',
            'rdfa1.1 = pyRdfa.rdflibparsers:RDFaParser',
            'application/svg+xml = pyRdfa.rdflibparsers:RDFaParser',
            'application/xhtml+xml = pyRdfa.rdflibparsers:RDFaParser',
            'hturtle = pyRdfa.rdflibparsers:HTurtleParser' # missing trailing comma
            'html = pyRdfa.rdflibparsers:StructuredDataParser'
            'html/text = pyRdfa.rdflibparsers:StructuredDataParser' # correct mime type is text/html
        ],
```
I wasn't sure what the deal is with `html/text`, so I just copied that line, changed the name to `text/html`, and added the missing commas. I reinstalled and ran the code snippet again, then got an error: `NameError: name 'warnings' is not defined`. I addressed that by adding an import of warnings not behind a conditional. I added a dependency on "requests" to setup.py and reinstalled. I got a NameError on ExecutionContext in rdflibparsers, so added an import. Then, I stopped because I figured someone else may already have done this who actually uses pyrdfa3 but who hasn't gotten it merged to master yet :)

**Reply by ghost** [reply]
Okay, thanks for the detail, most useful, I'd forgotten about the requirement for `requests`.

Actually, splitting it out as a separate plugin is fairly trivial, given that RDFLib has a plugin interface. The essential additions are few (adding the plugin registration to `plugin.py` plus adding the parsers to `plugins/parsers`) and are detailed in [this commit](https://github.com/gjhiggins/rdflib/commit/9ad094e2278b1abf7e9d41bb5e5406820be0a8e6).

In a split-off plugin, none of those changes to RDFLib would persist, they'd all be hived off to the separate plugin code and the registration would be handled by `setup.py` in exactly the same way that the [registration](https://github.com/RDFLib/rdflib-jsonld/blob/2d88ef989f89db8700d696a84086568879828753/setup.py#L105) of the parser/serializer for json-ld is handled.

So basically, the plugin code would consist of the two parsers and a setup.py which installs pyrdfa3 from PyPi and pymicrodata from git (happily, there's an [example in rdflib-leveldb](https://github.com/RDFLib/rdflib-leveldb/blob/a0f3386c71e6b1cfbd09c257e29400f5fde43ed3/setup.py#L16)) and that'd be about it. Oh, and some tests, ofc but they'd be contained in the plugin repos. 

I'll re-create it as an RDFLib plugin, commit it to a repos in my personal github space and then post a notification that it's available for tyre-kicking.

**Reply by ghost** [reply]
> I'll re-create it as an RDFLib plugin, commit it to a repos in my personal github space and then post a notification that it's available for tyre-kicking.

Done. Repos is [here](https://github.com/gjhiggins/rdflib-rdfa-microdata). The deprecation warnings from pyrdfa3 can safely be ignored.

**Comment by ghost** [comment]
And, on reflection, the branch could be fairly straightforwardly reworked as a plugin, using the same approach as the plugin stores.

Despite the branch name, I _wasn't_ thinking of proposing it be returned to RDFLib core, I was just checking the feasibility of getting it working at all.

Documentation needs work and it could use some better/more extensive tests.

**Reply by ghost** [reply]
> Despite the branch name, I _wasn't_ thinking of proposing it be returned to RDFLib core, I was just checking the feasibility of getting it working at all.

fwiw, more reflections on strategy/tactics in the draft PR https://github.com/RDFLib/rdflib/pull/1503
